### PR TITLE
GitHub Issue NOAA-EMC/GSI#282. GSL enhancements to the non-var cloud analysis.

### LINKED ISSUE #282: Enhance non-var cloud analysis
In past several years, several improvements to the non-var cloud analysis have been made by GSL and EMC for RAP/HRRR, 3DRTMA, and RRFS:

1) more options to decide how to do T/Q adjustment based on the cloud information.
2) check and adjust Qr/qs/qg and Qnr for each vertical profile to reduce the background reflectivity ghost in final analysis
3) Add function to build a vertical depended cloud range for METAR cloud observations. 

More options to build q innovation based on the cloud information. This is not part of non-var cloud analysis but is closely related.

### PR DESCRIPTION
**Pull request by hu5970**
This PR address the issue #282:
1) GSL enhancements to the non-var cloud analysis.
2) Update the code generating pseudo moisture observations from cloud observations.

### CODE REVIEW

**Review by ShunLiu-NOAA — COMMENTED** [review]
Ming,

There are a lot of changes in this PR. My comments are here:
1. Since this PR includes many parameter changes, are these change already tested in any parallel experiment?
2. src/gsi/read_Lightning.f90:
    line 424 to 428, is it possible avoid "stop2(50)", it might potentially cause the operation failure. 
3. src/gsi/reorg_metar_cloud.f90:
Could you please remove the line 284 to 288, line 311 and 312?
Also, is it possible to not use stop2(50) in line 322?

Shun

**Comment by hu5970** [comment]
@ShunLiu-NOAA 

Thanks for reviewing the code:

1)  There are four functions are added (listed in issue 282):
          a. more options to decide how to do T/Q adjustment based on the cloud information.
               Tested with retro and cases
          b. check and adjust Qr/qs/qg and Qnr for each vertical profile to reduce the background reflectivity ghost in final analysis
               This is used in the 3DRTMA parallel tests
         c.Add function to build a vertical depended cloud range for METAR cloud observations.
               This is tested with cases. We did not run parallel test for it and will do it for RRFS.
         d. More options to build q innovation based on the cloud information. This is not part of non-var cloud analysis but is closely related.
                This is tested with cases and retros.

2) stop(50) in read_lightning and reorg_metar_cloud will crash the GSI but it is used to give clear crash point and message when GSI fills data out of allocated memory. We have a way to avoid this by counting observations before allocate memory. There are many GSI observation IO use the current way to read the observations and current method does not cause trouble for RAP/HRRR operation.

3) will clean the code

File: src/gsi/reorg_metar_cloud.f90 [hunk 1]
```diff
@@ -260,19 +274,62 @@ subroutine reorg_metar_cloud(cdata,nreal,ndata,cdata_all,maxobs,ngrid)
...
+                         endif
+                      enddo
+                      if(ic==0) then
+                         if(abs(cdata_temp(6)) < 0.0001_r_kind) ic=1  ! clear
+                      else
+                         do k=1,6
+                            cdata_temp(5+k)=cloudlevel_temp(0+k)
+                            cdata_temp(11+k)=cloudlevel_temp(6+k)
+                         enddo
+!                      write(*,'(10f10.1)') (cdata_temp(5+k),k=1,6)
```
Comment (GangZhao-NOAA) [inline src/gsi/reorg_metar_cloud.f90:311]: Hi Ming,

For the algorithm and code, I do not have any comment. 
Just do not understand the usage of "cdata_temp" and "cloudlevel_temp".
From line 278 to line 312, the cdata_temp and cloudlevel_temp are modified and updated, but not output or used. So I am kind of confused. 
Anyway, as for the algorithm and code, I have no comment.

Thank you!
- Gang

**Review by GangZhao-NOAA — APPROVED** [review]
Hi Ming (@hu5970) ,

I have no comments and approve this PR. Just one comment for the "src/gsi/reorg_metar_cloud.f90". The new variables "cdata_temp" and "cloudlevel_temp" are updated but not output or used finally. I do not know why.  Are they just used for code testing? I also add a single comment in that code.

In summary, I approve this merge.

Thank you!

- Gang

**Review by ShunLiu-NOAA — APPROVED** [review]

File: src/gsi/reorg_metar_cloud.f90 [hunk 2]
```diff
@@ -260,19 +274,54 @@ subroutine reorg_metar_cloud(cdata,nreal,ndata,cdata_all,maxobs,ngrid)
...
+                   endif
+
+                   if(ic > 0)  then
+                      iout = iout + 1
+                      if(iout > maxobs) then
+                         write(6,*)'reorg_metar_cloud:  ***Error*** ndata > maxobs '
+                         call stop2(50)
+                      end if
+                      do k=1,nreal
+                         cdata_all(k,iout) = cdata_temp(k,ista_min)
```
Comment (MichaelLueken-NOAA) [inline src/gsi/reorg_metar_cloud.f90:317]: While compiling, CMake failed with the following error message:

```
[ 45%] Building Fortran object src/gsi/CMakeFiles/gsilib_wrf.dir/reorg_metar_cloud.f90.o
/gpfs/dell2/emc/modeling/noscrub/Michael.Lueken/update/src/gsi/reorg_metar_cloud.f90(317): error #6351: The number of subscripts is incorrect.   [CDATA_TEMP]
                         cdata_all(k,iout) = cdata_temp(k,ista_min)
---------------------------------------------^
```

cdata_temp should only have a single subscript, nreal.  Please correct this issue.

**Review by MichaelLueken — COMMENTED** [review]
Hi Ming,

While attempting to compile these changes, CMake failed due to an incorrect number of subscripts for cdata_temp at line 317.  Please correct this and please add (i_kind) to the integer declaration.

File: src/gsi/reorg_metar_cloud.f90 [hunk 3]
```diff
@@ -84,10 +90,18 @@ subroutine reorg_metar_cloud(cdata,nreal,ndata,cdata_all,maxobs,ngrid)
   real(r_kind)    ::     spval_p
   parameter (spval_p = 99999._r_kind)
 
+  real(r_kind)    ::     mindx,dxij,delat_radius,delta_height,radiusij
+  integer         ::     isprdij
```
[the remark this answers is not in the document]
Comment (hu5970) [reply src/gsi/reorg_metar_cloud.f90:94]: Done

File: src/gsi/reorg_metar_cloud.f90 (hunk 2 above)
Comment (hu5970) [reply src/gsi/reorg_metar_cloud.f90:317]: Corrected

**Comment by hu5970** [comment]
@MichaelLueken-NOAA : Mike, I have fixed the bug and combined the commit logs into one following the instructions from PR 285. Thanks, Ming

**Review by MichaelLueken — COMMENTED** [review]
I have compiled the changes in debug mode and there was a variable in reorg_metar_cloud.f90 that was found to be unused.  Please remove the unused variable (noted in the review).  There are also a few questions related to cleaning up commented out lines that were added in the code review.  If the commented out lines won't be used in the future, please remove them (these have been noted in the review).  Finally, there is a line in gsdcloudanalysis.F90 that exceeds the 12 character limit (due to the comment on the line).  Please move the comment to the line before or after the current line to correct this issue (the line has been noted in the review).  The debug tests were run and all successfully ran through to completion.

The debug tests were run and all successfully ran through to completion.

While reviewing the changes, I noted that both this PR and PR #285 have the identical change to ush/build.comgsi.  Since PR #285 is currently out to the review committee, that will be going in first.  It is possible that the same change being in ush/build.comgsi will cause conflicts when that work is merged to the authoritative repository.  You might want to remove the ush/build.comgsi change from this PR to ensure that I can submit this work to the review committee once the first PR is merged.

File: src/gsi/reorg_metar_cloud.f90 [hunk 4]
```diff
@@ -42,8 +42,12 @@ subroutine reorg_metar_cloud(cdata,nreal,ndata,cdata_all,maxobs,ngrid)
 
   use kinds, only: r_kind,i_kind,r_double
   use gridmod, only: nlon,nlat
-  use constants, only: one,half
+  use gridmod, only: region_dy,region_dx
```
[the remark this answers is not in the document]
Comment (hu5970) [reply src/gsi/reorg_metar_cloud.f90:45]: done

File: src/gsi/gsdcloudanalysis.F90 [hunk 5]
```diff
@@ -907,6 +927,103 @@ subroutine  gsdcloudanalysis(mype)
         end do
         end do
      end do
+ 
+!    ---- verical check and adjustment to the analysis of precipitation
+!         in order to remove/reduce the backround reflectivity "ghost" in
+!         analysis.
+!         Note: here rain_3d, snow_3d have been already changed into unit of kg/kg.
+!     if(l_precip_vertical_check) then
```
[the remark this answers is not in the document]
Comment (hu5970) [reply src/gsi/gsdcloudanalysis.F90:935]: Line 935 is removed.

File: src/gsi/gsdcloudanalysis.F90 [hunk 6]
```diff
@@ -907,6 +927,103 @@ subroutine  gsdcloudanalysis(mype)
...
+                       max_retrieved_qrqs=snow_3d(i,j,imaxlvl_ref)+rain_3d(i,j,imaxlvl_ref)
+                       k_cap=min(imaxlvl_ref,nsig)
+                       do k=k_cap,1,-1
+                          if( ref_mos_3d(i,j,k) <= -100.0_r_kind ) then   !  dbz-obs-missing level
+                             qrqs_retrieved=snow_3d(i,j,k)+rain_3d(i,j,k)
+                             if(qrqs_retrieved > max_retrieved_qrqs .and. qrqs_retrieved > 0.0001_r_kind) then
+                                ratio_hyd_bk2obs=max(min(max_retrieved_qrqs/qrqs_retrieved,1.0_r_kind),0.0_r_kind)
+                                if(rain_3d(i,j,k) > zero) then
+                                   rain_3d(i,j,k) = rain_3d(i,j,k)*ratio_hyd_bk2obs
+                                   nrain_3d(i,j,k)= min(nrain_3d(i,j,k)/ratio_hyd_bk2obs*2.5_r_kind,qnr_limit)    ! 2.5(old) or 1.0(new4) or 1.5(new5/6) 2.5(old, new7) 2.0(new8)
```
[the remark this answers is not in the document]
Comment (hu5970) [reply src/gsi/gsdcloudanalysis.F90:998]: Done

File: src/gsi/gsdcloudlib_pseudoq_mod.f90 [hunk 7]
```diff
@@ -133,13 +136,15 @@ SUBROUTINE cloudCover_Surface_col(mype,nsig,&
 !====================================================================
 !  Begin
 !
+   !write(6,*) 'cloudCover_Surface', mype, i_cloud_q_innovation
```
[the remark this answers is not in the document]
Comment (hu5970) [reply src/gsi/gsdcloudlib_pseudoq_mod.f90:139]: removed

File: src/gsi/setupcldtot.F90 [hunk 8]
```diff
@@ -577,13 +579,16 @@ subroutine setupcldtot(obsLL,odiagLL,lunin,mype,bwork,awork,nele,nobs,is,conv_di
...
+                       error=one/(cloudqvis*3.E-01_r_kind)
                    else
                        cycle
                    endif
        
                    q_obcount=q_obcount+1
        
-                   error=one/(cloudqvis*3.E-01_r_kind)
+                   ! all obs errors = 30%
+                   !error=one/(cloudqvis*3.E-01_r_kind)
```
[the remark this answers is not in the document]
Comment (hu5970) [reply src/gsi/setupcldtot.F90:591]: removed

File: src/gsi/setupcldtot.F90 [hunk 9]
```diff
@@ -711,7 +716,8 @@ subroutine setupcldtot(obsLL,odiagLL,lunin,mype,bwork,awork,nele,nobs,is,conv_di
 
   !! Write information to diagnostic file
   if(conv_diagsave)then
-     if (i_cloud_q_innovation == 2 .and. iip>0) then
+     if (i_cloud_q_innovation >= 20 .and. iip>0) then
+!         call dtime_show(myname,'diagsave:q',i_q_ob_type)
```
[the remark this answers is not in the document]
Comment (hu5970) [reply src/gsi/setupcldtot.F90:720]: removed

**Comment by hu5970** [comment]
@MichaelLueken-NOAA Mike, I have revised the code based on your comments. Please check. Thanks,

**Comment by MichaelLueken-NOAA** [comment]
@hu5970 After merging PR #285 to the authoritative repo, there is now a conflict in src/gsi/gsimod.F90.  Please merge the latest NOAA-EMC/GSI master into your fork's gsl_nonvar_cloudanalysis branch using:

git remote add upstream https://github.com/NOAA-EMC/GSI.git
git remote update
git rebase upstream/master
Correct the comment conflict in src/gsi/gsimod.F90
git add src/gsi/gsimod.F90
git rebase --continue
git push origin gsl_nonvar_cloudanalysis --force

Once this is complete, I should be able to send this work out to the review committee.  Please let me know if you have any questions or encounter any issues.

**Comment by hu5970** [comment]
@MichaelLueken-NOAA I have solved conflict and rebased the branch. Thanks.

**Comment by MichaelLueken-NOAA** [comment]
The due date for feedback from the DA review committee has passed.  I will now give final approval to these changes and merge them to the authoritative repository.

**Pull request merged**